Toggle containers programmatically

petrsimon opened the pull request:
Simply add `collapsed` attribute to the `ui-layout-container` you want to control and toggle the value.
```javascript
$scope.layout = {
    toolbar: false,
    statusbar: true
}
```

```html
<div ui-layout-container data-collapsed="layout.toolbar">TOOLBAR</div>
<div ui-layout-container data-collapsed="layout.statusbar">STATUSBAR</div>
```

BobbieBarker commented:
Going to need a more extensive code review tomarrow. This pr needs unit tests as well.

BobbieBarker commented:
Also the commit history needs cleaned up

petrsimon commented:
Cleaned up the commits. Please, have a look and let me know if this is a way to go. I will then write the unit tests.

SomeKittens commented:
Thanks, @petrsimon!  This is an oft-requested feature.  I'll let @BobbieBarker do some review, then tests, then bring this in!  Glad to have it.

SomeKittens commented on src/ui-layout.js:
```diff
@@ -41,7 +41,8 @@ angular.module('ui.layout', [])
     opts.minSizes = opts.minSizes || [];
-    opts.dividerSize = opts.dividerSize || 10; //default divider size set to 10
+
+    opts.dividerSize = opts.dividerSize === undefined ? 10 : opts.dividerSize; //default divider size set to 10
```
Is this related to this commit in particular?  If not, it should be part of a separate PR.

petrsimon replied:
You are right, not really related. I forgot to take that out. Will PR this separately.

BobbieBarker commented on src/ui-layout.js:
```diff
@@ -796,6 +807,31 @@ angular.module('ui.layout', [])
+                  angular.element(el).trigger('click');
+                });
+
+                $timeout(function(){
```
In this context it looks like $timeout is being used to handle async control flow, would it be possible to rewrite this using a promise/$q. 
Other than that the PR looks good to me.

wesleycho commented on src/ui-layout.js:
```diff
@@ -796,6 +807,37 @@ angular.module('ui.layout', [])
             if(!element.hasClass('ui-layout-container')) element.addClass('ui-layout-container');
 
+            scope.$watch('collapsed', function (val, old) {
+              if (old !== undefined && val !== old) {
```
This is better as `if (angular.isDefined(old)) {` - `val` is guaranteed to not equal `old` by how $watch works.

petrsimon replied:
Actually during initialization, the equality can actually happen, which would in turn toggle the container against expectations.

wesleycho commented on src/ui-layout.js:
```diff
@@ -796,6 +807,37 @@ angular.module('ui.layout', [])
+
+
+                $timeout(function(){
+                  angular.element(el).trigger('click');
```
This looks like a hack...

petrsimon replied:
Does it ;). I've improved this in the new commit.

petrsimon commented:
I've made few improvements, please see the new commit.

BobbieBarker commented on src/ui-layout.js:
```diff
@@ -796,6 +807,37 @@ angular.module('ui.layout', [])
+                }
+
+
+                $timeout(function(){
```
I hate to be a stickler but I was thinking more about using $q instead of $timeout. In my opinion $timeout isn't really appropriate here.

petrsimon replied:
Aha. This is the trick I know how to get around "$digest in progess" error, which was thrown after trying to trigger the click. Not sure how to use $q here.

petrsimon replied:
Btw, why it $timeout not appropriate. I thought this is a common way how to get around $digest error, to execute a routine in another digest cycle with $timeout.

petrsimon commented:
I've cleaned up few bits, improved it a bit and added few test and demo. Please check it and let me know how it looks now. Will squash the commits later when it's ready for pull.

SomeKittens commented on src/ui-layout.js:
```diff
@@ -239,14 +240,14 @@ angular.module('ui.layout', [])
-            //opts.collapsed[i] = child.attr('collapsed') || opts.collapsed[i] || false;
-
 
+            opts.sizes[i] = child.attr('size') || child.attr('data-size') || opts.sizes[i];
```
Why drop `auto` as an option?

petrsimon replied:
Because the next line takes care of it in case there's no size provided by the user. Am I missing something?

petrsimon commented:
I've ran into a resizing issues with collapsed containers, which lead me to https://github.com/petrsimon/ui-layout/tree/refactor-display-update (also announced in #100, which it is solution to. Will wait for comments, especially @j--w who was working on it).
I've already merged that branch with this one in https://github.com/petrsimon/ui-layout/tree/toggle-and-resize, which seem to complement each other quite nicely. 
Not sure if I should push it over this one or create new PR from `toggle-and-resize?

petrsimon commented:
I've made quite few improvements in the toggle-and-resize branch and it all seems to be behaving quite nicely. I had to cut a bit deeper to get around bugs that appeared and I'm not sure if I should force push it here or rather create new PR. Please see b4eb6c2. I will cool off for a while and stop working on it, unless I run into bugs in my application.

j--w commented:
@petrsimon I made some comments on #100. I would think that tidying up the toggle-and-resize branch and creating a new PR would be the best course of action.

petrsimon commented:
@j--w, thanks. I thought new PR might be better. Closing this in favour of #154.
